### DISCUSSION TITLE
Support Dynamic Island and Live Activities

### DISCUSSION THREAD

**Discussion by FledgeXu** [kiwix/apple#493] [Ideas]
In iOS 16 and iPhone 14, Apple brings [Dynamic island](https://developer.apple.com/documentation/widgetkit/dynamicisland) and [Live Activities](https://developer.apple.com/documentation/activitykit/displaying-live-data-with-live-activities). 

Since kiwix needs often to download big files, it would be nice that use those two features to show the progress of downloading.

**Comment by kelson42** [comment]
@FledgeXu This is a technical, solution driven, description, please explain:
* what is the problem
* how should it looks like, once implwmented

**Comment by FledgeXu** [comment]
@kelson42 
For now, this is the only way to see the progress of the download:
![IMG_0580](https://user-images.githubusercontent.com/16664654/227201693-94009138-b586-429d-8dd2-789dee5fd5e0.PNG)
When you want to download a large zim file, it's annoying to have to repeat this process of checking progress several times.

Once we add the Dynamic island and Live Activities support, we can show this progress outside the app, so users don't need to open the app and check the progress of the download anymore.

like this:

(iPhone 14 Pro or later)
![image](https://user-images.githubusercontent.com/16664654/227202430-fd08ec36-f2ea-40fe-8f0d-bc4e24f5b742.png)

(All users)
<img width="317" alt="image" src="https://user-images.githubusercontent.com/16664654/227204012-5ae97f2a-1661-4895-a0fd-f12958a3cfd8.png">

**Comment by automactic** [comment]
This is a cool idea, I was looking into the same thing earlier, and I bet it would be very fun to use if I can get it working. However this is not currently possible within current constraint of Live Activities.

From Apple's developer documentation site [[link](https://developer.apple.com/documentation/activitykit/displaying-live-data-with-live-activities#Understand-constraints)]:

> Each Live Activity runs in its own sandbox, and — unlike a widget — it can’t access the network or receive location updates. To update the dynamic data of an active Live Activity, use the ActivityKit framework in your app or allow your Live Activities to receive ActivityKit push notifications as described in [Updating and ending your Live Activity with ActivityKit push notifications](https://developer.apple.com/documentation/activitykit/updating-and-ending-your-live-activity-with-activitykit-push-notifications).

Live activities cannot access network (which means I cannot query the network interface for download task progress), can be updated only via server side push notification (which does not make sense for action happening on device), or via `ActivityKit` (which means the app has to be running)

It seems like "timer" live activities can update every second, but file download progress cannot be simply implemented as a timer.

We'll have to see if this change in WWDC 23 / iOS 17, but sorry there isn't anything I can do right now

**Comment by automactic** [comment]
I am moving this to Discussion -> Idea, because this is an interesting idea, but nothing to be done right now
